modules/services/yggdrasil.nix

```diff
@@ -0,0 +1,117 @@
+            "tcp://[aaaa:bbbb:cccc:dddd::eeee]:fffff"
+          ];
+          Listen = [
+            "tcp://0.0.0.0:xxxxx"
+          ];
+        };
+        description = ''
+          Configuration for yggdrasil, as a Nix attribute set.
+
+          Warning: this is stored in the WORLD-READABLE Nix store!
+          Therefore, it is not appropriate for private keys. If you
+          wish to specify the keys, use {option}`configFile`.
+
+          If no keys are specified then ephemeral keys are generated
+          and the Yggdrasil interface will have a random IPv6 address
+          each time the service is started. This is the default.
+
+          If both {option}`configFile` and {option}`settings`
+          are supplied, they will be combined, with values from
+          {option}`configFile` taking precedence.
+
+          You can use the command `nix-shell -p yggdrasil --run "yggdrasil -genconf"`
+          to generate default configuration values with documentation.
+        '';
+      };
+
+      configFile = mkOption {
+        type = nullOr path;
+        default = null;
+        example = "/run/keys/yggdrasil.conf";
+        description = lib.mdDoc ''
+          A file which contains JSON or HJSON configuration for yggdrasil. See
+          the {option}`settings` option for more information.
+
+          On NixOS, file in this option is limited to 1 MB due to limitations 
+          in systemd. If you would like to share your yggdrasil configuration
+          between nix-darwin and NixOS, you should keep this limitation in mind,
+          even though there is no equivalent limit on macOS.
+        '';
+      };
+
+      package = mkPackageOption pkgs "yggdrasil" { };
+
+      extraArgs = mkOption {
+        type = listOf str;
+        default = [ ];
+        example = [ "-loglevel" "info" ];
+        description = lib.mdDoc "Extra command line arguments.";
+      };
+    };
+  };
+
+  config = mkIf cfg.enable (
+    let
+      yggdrasilConf = "/run/yggdrasil/yggdrasil.conf";
+      binYggdrasil = "${cfg.package}/bin/yggdrasil";
+      binHjson = "${pkgs.hjson-go}/bin/hjson-cli";
+    in
+    {
+      environment.systemPackages = [ cfg.package ];
+
+      # have to write it in that way to not interfere with brew's (or idk github?) ygg.plist
+      launchd.daemons.ygg =
+        {
+          script = ''
+            set -euo pipefail
+
+            mkdir -p $(dirname ${yggdrasilConf})
+
+            # prepare config file
+            ${(if settingsProvided || configFileProvided then
+              "echo "
+
+              + (lib.optionalString settingsProvided
+                "'${builtins.toJSON cfg.settings}'")
+              + (lib.optionalString configFileProvided
+                "$(${binHjson} -c ${cfg.configFile})")
+              + " | ${pkgs.jq}/bin/jq -s add | ${binYggdrasil} -normaliseconf -useconf"
+            else
+              "${binYggdrasil} -genconf") + " > ${yggdrasilConf}"}
```

Comment:
What happens if the config file already exists?

Reply:
Thanks for pointing this!

In the upstream ygg module this was expected behavior since it has `persistentKeys` option.

I changed this to persist config between service restarts, if it already exists.